libbeat/cfgfile/reload.go

```diff
@@ -15,3 +31,158 @@ type ReloadConfig struct {
+func NewReloader(cfg *common.Config) *Reloader {
+
+	config := DefaultReloadConfig
+	cfg.Unpack(&config)
+
+	return &Reloader{
+		registry: NewRegistry(),
+		config:   config,
+		done:     make(chan struct{}),
+	}
+}
+
+// Run runs the reloader
+func (r *Reloader) Run(wrapper Wrapper) {
+
+	logp.Info("Config reloader started")
+
+	r.wg.Add(1)
+	defer r.wg.Done()
+
+	// Stop all running modules when method finishes
+	defer r.stopModules(r.registry.CopyList())
+
+	path := r.config.Path
+	if !filepath.IsAbs(path) {
+		path = paths.Resolve(paths.Config, path)
+	}
+
+	gw := NewGlobWatcher(path)
+
+	for {
+		select {
+		case <-r.done:
+			logp.Info("Dynamic config reloader stopped")
+			return
+		case <-time.After(r.config.Period):
+
+			debugf("Scan for new config files")
+			configReloads.Add(1)
+
+			files, updated, err := gw.Scan()
+			if err != nil {
+				// In most cases of error, updated == false, so will continue
+				// to next iteration below
+				logp.Err("Error fetching new config files: %v", err)
+			}
+
+			// no file changes
+			if !updated {
+				continue
+			}
+
+			// Load all config objects
+			configs := []*common.Config{}
+			for _, file := range files {
+				c, err := LoadList(file)
+				if err != nil {
+					logp.Err("Error loading config: %s", err)
+					continue
+				}
+
+				configs = append(configs, c...)
+			}
+
+			debugf("Number of module configs found: %v", len(configs))
+
+			startList := map[uint64]Runner{}
+			stopList := r.registry.CopyList()
+
+			for _, c := range configs {
+
+				// Only add prospectors to startlist which are enabled
+				if !c.Enabled() {
+					continue
+				}
+
+				w, hash, err := wrapper.GetWrapper(c)
+				if err != nil {
+					logp.Err("Error creating module wrapper: %s", err)
+					continue
+				}
+
+				debugf("Remove module from stoplist: %v", hash)
+				delete(stopList, hash)
+
+				// As module already exist, it must be removed from the stop list and not started
+				if !r.registry.Has(hash) {
+					debugf("Add module to startlist: %v", hash)
+					startList[hash] = w
+					continue
+				}
+			}
+
+			r.stopModules(stopList)
+			r.startModules(startList)
+		}
+	}
+}
+
+// Stop stops the reloader and waits for all modules to properly stop
+func (r *Reloader) Stop() {
+	close(r.done)
+	r.wg.Wait()
+}
+
+func (r *Reloader) startModules(list map[uint64]Runner) {
+
+	logp.Info("Starting %v modules ...", len(list))
+	for hash, mw := range list {
+		//mr := NewRunner(r.client, mw)
+		mw.Start()
+		r.registry.Add(hash, mw)
+		moduleStarts.Add(1)
+		moduleRunning.Add(1)
+		debugf("New Module Started: %v", hash)
+	}
+}
+
+func (r *Reloader) stopModules(list map[uint64]Runner) {
+	logp.Info("Stopping %v modules ...", len(list))
+
+	wg := sync.WaitGroup{}
+	for hash, w := range list {
+		wg.Add(1)
+		// Stop modules in parallel
+		func() {
+			defer wg.Done()
+			w.Stop()
+			r.registry.Remove(hash)
+			moduleStops.Add(1)
+			moduleRunning.Add(-1)
+			debugf("Module stopped: %v", hash)
+		}()
```

Comment:
Is there a test for this? `hash` and `w` should be passed into the function through arguments.